### DISCUSSION TITLE
Some questions about throat rate

### DISCUSSION THREAD

**Discussion by Devin-one** [PMEAL/OpenPNM#2699] [Q&A]
Hello, professor. I use `alg.rate(throats=pn.Ts, rate='single') `to obtain the rate of the longitudinal throat and the rate of the transverse throat. But their distribution patterns are very different. Is the vector sum of these two rates the real rate? Thank you in advance for your reply.

**Comment by jgostick** [comment]
No the throat rates are all independent. The balances are done around the pores, so the rate through each pore is zero except for boundary pores or if there is a reaction.  If you sum up all the throat rates you'll count the same mass multiple times.  One other gotcha with throats is that the direction is arbitrary, from pore i to j, where i and j refer to the pore indices in the conns array.

**Reply by Devin-one** [reply]
Thank you for your patience. I still have some confusion. What I build is a two-dimensional network. As far as I know, the number of throat follows the rule of first longitudinal and then transverse. So I guess the rate array obtained by this method, where the first half of the array is the rate of the longitudinal throat, and the rest is the rate of the transverse throat. Is my guess right? Also, the obtained throat rate is positive and how to distinguish inflow and outflow?

**Reply by jgostick** [reply]
I wouldn't rely on the throats having any order.  The 'direction' will depend on the transport mechanism but is almost always from the pore with the highest driving force to that with the lower.

**Comment by Devin-one** [comment]
I saw the following figure in the tutorial. Therefore, I speculate that the array of throat rate obtained by `alg.rate(throats=pn.Ts, rate='single')` ,like pores, stores data in order. If not, how can we show the rate distribution of the entire network?
<img width="320" alt="2f7730465dde66374b7a4e280692e24" src="https://user-images.githubusercontent.com/93040662/220240600-da033063-3c31-40f5-92d9-826cde234d5a.png">

**Reply by jgostick** [reply]
Yes, rate does return the values in order as you see here.  Where I meant was that the order above will not be reliable if you use a random network or extraction.

To see the rate distribution you can probably just take the absolute value, and use that to color the throats, like:

``op.visualization.plot_connections(network=pn, throats=pn.Ts, color_by=np.abs(alg.rate(throats=pn.Ts)), mode='single)``

**Comment by Devin-one** [comment]
I tried the way you told me. But the code reported an error. 
![image](https://user-images.githubusercontent.com/93040662/223370685-fa646337-e202-4469-9972-9a6efd767d77.png)
Here are all the modules I call.
```
import numpy as np
import matplotlib.pyplot as plt
import openpnm as op
import scipy.stats
import pandas as pd
from scipy.interpolate import Rbf
import tkinter
```
But the code can run normally when I use `ax= op.visualization.plot_connections(network=pn, color_by=sf.rate(throats=pn.Ts, mode='single'))`. I'm sorry to bother you again.

**Reply by jgostick** [reply]
There was a typo in my code, it should have been ``visualization.plot_connections``.  It's updated now.

**Reply by Devin-one** [reply]
Thank you for your timely reply. I tried and failed again. I wonder if it is because I built a two-dimensional network? I'm sorry I don't know much about Python.
![image](https://user-images.githubusercontent.com/93040662/223696633-bfc1ef97-a6a0-48a7-8a25-1f1733f927d9.png)
